src/components/abstractComponent.ts

```diff
@@ -0,0 +1,579 @@
+   * Returns false by default on the base Component class.
+   */
+  fixedHeight(): boolean;
+  /**
+   * Detaches a Component from the DOM. The Component can be reused.
+   *
+   * This should only be used if you plan on reusing the calling Component. Otherwise, use destroy().
+   *
+   * @returns The calling Component.
+   */
+  detach(): this;
+  /**
+   * Adds a callback to be called when the Component is detach()-ed.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {IComponent} The calling Component.
+   */
+  onDetach(callback: GenericComponentCallback<D>): this;
+  /**
+   * Removes a callback to be called when the Component is detach()-ed.
+   * The callback is identified by reference equality.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {IComponent} The calling Component.
+   */
+  offDetach(callback: GenericComponentCallback<D>): this;
+  /**
+   * Gets the parent ComponentContainer for this Component.
+   */
+  parent(): ComponentContainer;
+  /**
+   * Sets the parent ComponentContainer for this Component.
+   * An error will be thrown if the parent does not contain this Component.
+   * Adding a Component to a ComponentContainer should be done
+   * using the appropriate method on the ComponentContainer.
+   */
+  parent(parent: ComponentContainer): this;
+  parent(parent?: ComponentContainer): any;
+  /**
+   * @returns {Bounds} for the component in pixel space, where the topLeft
+   * represents the component's minimum x and y values and the bottomRight represents
+   * the component's maximum x and y values.
+   */
+  bounds(): Bounds;
+  /**
+   * Removes a Component from the DOM and disconnects all listeners.
+   */
+  destroy(): void;
+  /**
+   * Gets the width of the Component in pixels.
+   */
+  width(): number;
+  /**
+   * Gets the height of the Component in pixels.
+   */
+  height(): number;
+  /**
+   * Gets the origin of the Component relative to its parent.
+   *
+   * @return {Point}
+   */
+  origin(): Point;
+  /**
+   * Gets the origin of the Component relative to the root
+   *
+   * @return {Point}
+   */
+  originToRoot(): Point;
+  /**
+   * Gets the container holding the visual elements of the Component.
+   *
+   * Will return undefined if the Component has not been anchored.
+   *
+   * @return {IContent} content selection for the Component
+   */
+  content(): IContent;
+  /**
+   * Gets the top-level element of the component
+   */
+  element(): IContent;
+  /**
+   * Gets the root component of the hierarchy
+   *
+   * @return {IComponent}
+   */
+  root(): IComponent<any>;
+  /**
+   * Gets the translator for the component. A translator
+   * converts client coordinates to component coordinates and
+   * takes into account any transformations that have been applied
+   * to the element.
+   *
+   * @return {Utils.ITranslator}
+   */
+   translator(): Utils.Translator;
+}
+
+export abstract class AbstractComponent<D> implements IComponent<D> {
```

Comment:
Sure the rationale behind having the two is purely for sanitary reasons. I like separating definition from implementation as much as possible. 1) It makes for a more flexible API in the future (say you have some new crazy component with methods that you don't want in abstract component) at very little up front cost. 2) It keeps other APIs around the codebase a little more readable. Rather than typing out AbstractComponent<D> everywhere, you just type IComponent<D>. Additionally, at least in my mind, reading something with ICompnent<D> as a parameter or property presents slightly less cognitive strain than AbstractComponent<D>?